Add ActivityPub moderation system with hierarchical blocking

obenland opened the pull request:
See #418.

## Proposed changes:
* Add comprehensive ActivityPub moderation system with hierarchical blocking
* Implement three block types: actors (users), domains (instances), and keywords (content)
* Add site-wide moderation interface for admins in Settings → ActivityPub
* Add user-specific blocking controls in user profiles
* Update inbox controllers to filter activities based on moderation rules
* Create modern JavaScript interface with WordPress AJAX utilities
* Implement responsive UI matching WordPress design standards

### Other information:

- [ ] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

**Site-wide Moderation (Admin Users)**
* Go to 'Settings → ActivityPub'
* Scroll down to the "Moderation" section
* Test adding actors by full URL (e.g., https://mastodon.social/@username)
* Test adding domains (e.g., spam-instance.com)
* Test adding keywords (e.g., spam, advertisement)
* Verify all blocks are saved and displayed correctly
* Test removing blocks using the "Remove" buttons

**User-specific Blocking**
* Go to 'Users → Your Profile' (or edit another user's profile as admin)
* Scroll to the "Moderation" section under ActivityPub settings
* Test adding personal blocks for actors, domains, and keywords
* Verify these only affect the specific user, not site-wide
* Test removing personal blocks

**Activity Filtering**
* Send test ActivityPub activities that match blocked criteria
* Verify activities are filtered correctly at both levels
* Check that blocked activities trigger the `activitypub_rest_inbox_disallowed` action
* Verify hierarchical blocking: site-wide blocks affect all users, user blocks only affect individual users

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

<!-- Add a changelog message here -->

Comprehensive ActivityPub moderation system with hierarchical blocking for actors, domains, and keywords. Includes site-wide admin controls and individual user blocking options.

</details>

obenland committed:
Add comprehensive ActivityPub moderation system

Implements hierarchical blocking system (site-wide → user-specific → WordPress disallowed list) with three block types:
- Actors: Block specific ActivityPub users by URL
- Domains: Block entire instances by domain
- Keywords: Block content containing specific keywords

Features:
- Site-wide moderation for admins via Settings → ActivityPub
- User-specific blocks via user profiles
- Modern JavaScript with WordPress AJAX utilities
- Responsive UI matching WordPress standards
- Proper security with nonces and permission checks
- Integration with existing inbox controllers

obenland commented on includes/class-moderation.php:
```diff
@@ -0,0 +1,334 @@
+		return array(
+			'actors'   => \get_user_meta( $user_id, self::USER_BLOCKED_ACTORS_META, true ) ?: array(),
+			'domains'  => \get_user_meta( $user_id, self::USER_BLOCKED_DOMAINS_META, true ) ?: array(),
+			'keywords' => \get_user_meta( $user_id, self::USER_BLOCKED_KEYWORDS_META, true ) ?: array(),
```
@pfefferle Should actor blocks work in the same way as followers and following, with actor-specific post meta? Or like this, where we keep a list of actor IDs in user meta?

pfefferle replied:
good question!?!

It seems that the current version uses whatever you type in the input field, without a lookup. So I would suggest doing that lookup to store the real ID and not any kind of profile URL or WebFinger handle and then we can use the Actor to store the info directly.

pfefferle replied:
I think this serialized list might work for now, but it is very hard to maintain it when you have more than a hundred items and might want/need to page. Maybe we should think about a taxonomy or a CPT instead.

obenland replied:
Your feedback makes me think it should work like follow and following :) 

We could replicate the following UI and add the Block link to the other two tables, and remove the actors setting in favor of the blocked table.

pfefferle replied:
I think it is also fine to start with the system wide settings and add the user specific ones later on.

obenland committed:
Add comprehensive PHPUnit tests for Moderation API

Implements thorough test coverage for the ActivityPub moderation system including:

- User block management (add/remove/get) for all three types (actors, domains, keywords)
- Site-wide block management with admin controls
- Activity blocking logic with hierarchical priority testing
- Edge cases: malformed data, invalid inputs, special characters, Unicode
- Complex scenarios: duplicate blocks, array re-indexing, WordPress fallback
- Boundary testing: extremely long values, invalid user IDs, empty data
- Integration testing: WordPress comment disallowed list fallback

All 22 tests pass with 91 assertions covering core functionality,
error handling, and real-world usage scenarios.

obenland committed:
Apply PHPCBF code formatting fixes

Automated code style fixes applied by PHP_CodeSniffer including:
- Consistent spacing and indentation
- WordPress coding standards compliance
- Array formatting and alignment
- Comment formatting improvements

No functional changes, only code style improvements.

obenland committed:
Improve moderation JavaScript UX and fix terminology

- Remove page reloads from AJAX operations for better UX
- Replace location.reload() with dynamic UI updates
- Add helper functions to manage DOM updates properly
- Fix table removal when last item is deleted
- Rename functions from 'block' to 'blockedTerm' to avoid confusion with WordPress blocks
- Fix issue where adding terms created tables in all sections
- Use specific selectors to target correct containers for each type

obenland committed:
Add code coverage annotations and refactor blocking logic

- Add @covers annotations to all test methods for proper coverage tracking
- Add @coversDefaultClass to test and main Moderation classes
- Refactor activity_is_blocked_site_wide and activity_is_blocked_for_user to eliminate code duplication
- Extract common blocking logic into private check_activity_against_blocks method
- Maintain full test coverage with all 22 tests passing

Improves code maintainability and test coverage visibility.

obenland committed:
Add changelog entry for blocking and moderation system

Documents the new comprehensive moderation features including:
- User-specific blocking controls for actors, domains, and keywords
- Site-wide moderation interface for administrators
- Integration with existing WordPress disallowed list
- Modern responsive UI with WordPress utilities

pfefferle commented:
What about the Blog-User?

obenland commented:
The blog user would adhere to site-wide settings

obenland committed:
Fix AJAX nonce verification order for better security

Move nonce verification to the beginning of all AJAX handlers before accessing any  data, as recommended by Copilot security review. This ensures proper nonce validation occurs before any other processing.

- ajax_add_user_block: Verify nonce first
- ajax_remove_user_block: Verify nonce first
- ajax_add_site_block: Verify nonce first
- ajax_remove_site_block: Verify nonce first, remove duplicate check

All tests continue to pass (22 tests, 91 assertions).

pfefferle commented:
The Actor list is formatted differently as the other two.

obenland commented on includes/class-moderation.php:
```diff
@@ -0,0 +1,310 @@
+	 * @return bool True if blocked, false otherwise.
+	 */
+	public static function activity_is_blocked_for_user( $activity_data, $user_id ) {
+		$blocked_actors   = \get_user_meta( $user_id, self::USER_BLOCKED_ACTORS_META, true ) ?: array(); // phpcs:ignore Universal.Operators.DisallowShortTernary.Found
+		$blocked_domains  = \get_user_meta( $user_id, self::USER_BLOCKED_DOMAINS_META, true ) ?: array(); // phpcs:ignore Universal.Operators.DisallowShortTernary.Found
+		$blocked_keywords = \get_user_meta( $user_id, self::USER_BLOCKED_KEYWORDS_META, true ) ?: array(); // phpcs:ignore Universal.Operators.DisallowShortTernary.Found
```
This suggestion is incorrect. `get_user_meta()` returns `false` when no meta exists, not `null`. The behavior differs:

- `false ?? array()` would return `false` (because `false` is not `null`)
- `false ?: array()` correctly returns `array()` (because `false` is falsy)

The short ternary operator is the correct choice here and the phpcs:ignore is appropriate.

obenland commented on includes/wp-admin/class-moderation.php:
```diff
@@ -0,0 +1,176 @@
+	/**
+	 * AJAX handler to add user block.
+	 */
+	public static function ajax_add_user_block() {
```
Fixed in commit 72f919e8

You are absolutely right about the security concern. I have moved nonce verification to the beginning of all AJAX handlers before accessing any $_POST data:

- ajax_add_user_block(): Nonce check now first
- ajax_remove_user_block(): Nonce check now first  
- ajax_add_site_block(): Nonce check now first
- ajax_remove_site_block(): Nonce check now first

Thanks for catching this security issue\!

obenland commented on includes/wp-admin/class-moderation.php:
```diff
@@ -0,0 +1,176 @@
+	/**
+	 * AJAX handler to remove user block.
+	 */
+	public static function ajax_remove_user_block() {
```
Fixed in commit 72f919e8

Same security fix applied here - nonce verification now happens first before accessing any $_POST data. Thanks for the thorough security review\!

obenland commented on includes/class-moderation.php:
```diff
@@ -0,0 +1,310 @@
+			return false;
+		}
+
+		$blocks = \get_user_meta( $user_id, $meta_key, true ) ?: array(); // phpcs:ignore Universal.Operators.DisallowShortTernary.Found
```
I appreciate the suggestion, but I decided not to implement this helper function. The wp.a11y availability check is only used a few times, and the current approach is clear and follows WordPress patterns. Adding the complexity of a helper function doesn't justify the minor reduction in code duplication. The current implementation is maintainable and easily understood.

obenland replied:
This suggestion is incorrect. get_user_meta() returns false when no meta exists, not null. The behavior differs:

- false ?? array() would return false (because false is not null)
- false ?: array() correctly returns array() (because false is falsy)

The short ternary operator is the correct choice here and the phpcs:ignore is appropriate.

obenland commented on assets/js/activitypub-moderation-admin.js:
```diff
@@ -0,0 +1,260 @@
+
+			if ( ! value ) {
+				// Use wp.a11y.speak for better accessibility.
+				if ( wp.a11y && wp.a11y.speak ) {
+					wp.a11y.speak( activitypubModerationL10n.enterValue, 'assertive' );
+				} else {
+					alert( activitypubModerationL10n.enterValue );
+				}
```
I appreciate the suggestion, but I decided not to implement this helper function. The wp.a11y availability check is only used a few times, and the current approach is clear and follows WordPress patterns. Adding the complexity of a helper function doesn't justify the minor reduction in code duplication. The current implementation is maintainable and easily understood.

obenland commented on includes/class-moderation.php:
```diff
@@ -0,0 +1,310 @@
+			return false;
+		}
+
+		$blocks = \get_user_meta( $user_id, $meta_key, true ) ?: array(); // phpcs:ignore Universal.Operators.DisallowShortTernary.Found
```
This suggestion is incorrect. get_user_meta() returns false when no meta exists, not null. The behavior differs:

- false ?? array() would return false (because false is not null)
- false ?: array() correctly returns array() (because false is falsy)

The short ternary operator is the correct choice here and the phpcs:ignore is appropriate.

obenland commented on includes/class-moderation.php:
```diff
@@ -0,0 +1,310 @@
+		return array(
+			'actors'   => \get_user_meta( $user_id, self::USER_BLOCKED_ACTORS_META, true ) ?: array(), // phpcs:ignore Universal.Operators.DisallowShortTernary.Found
+			'domains'  => \get_user_meta( $user_id, self::USER_BLOCKED_DOMAINS_META, true ) ?: array(), // phpcs:ignore Universal.Operators.DisallowShortTernary.Found
+			'keywords' => \get_user_meta( $user_id, self::USER_BLOCKED_KEYWORDS_META, true ) ?: array(), // phpcs:ignore Universal.Operators.DisallowShortTernary.Found
```
This suggestion is incorrect. get_user_meta() returns false when no meta exists, not null. The behavior differs:

- false ?? array() would return false (because false is not null)
- false ?: array() correctly returns array() (because false is falsy)

The short ternary operator is the correct choice here and the phpcs:ignore is appropriate.

obenland commented on includes/wp-admin/class-settings-fields.php:
```diff
@@ -151,6 +160,33 @@ public static function register_settings_fields() {
 			array( 'label_for' => 'activitypub_relays' )
 		);
+
+		// Only show moderation fields to users with manage_options capability.
```
I appreciate the suggestion, but I believe the current comment is clear and concise. The existing comment effectively communicates the intent without being overly verbose. Adding a lengthy explanation about security implications doesn't add meaningful value when the capability check itself is self-explanatory. The current implementation follows standard WordPress patterns.

obenland commented:
> The Actor list is formatted differently as the other two.

Is that a good thing?

obenland committed:
Consolidate moderation admin functionality and fix AJAX parameter conflict

- Dissolve separate wp-admin/class-moderation.php into main class-admin.php
- Rename AJAX action to 'activitypub_moderation_settings' for clarity
- Fix AJAX parameter conflict by changing 'action' to 'operation'
- Remove separate moderation class initialization from main plugin file
- Integrate moderation scripts and handlers into unified Admin class
- Update JavaScript to use single nonce and new parameter names

The 'action' parameter was conflicting with WordPress's internal AJAX routing
mechanism, causing "Invalid context or action" errors. Using 'operation'
resolves this conflict.

Benefits:
- Reduces codebase by 70 lines while maintaining full functionality
- More descriptive AJAX action name clearly indicates purpose
- Simplified admin architecture with fewer files to maintain
- All 22 moderation tests continue to pass with 91 assertions

obenland committed:
Add moderation meta and settings for ActivityPub

Introduces user meta and site-wide settings for blocking ActivityPub actors, domains, and keywords. This enables more granular moderation capabilities for both individual users and the entire site.

obenland committed:
Align variable assignments in ajax_moderation_settings

Adjusted whitespace for variable assignments in the ajax_moderation_settings method to improve code readability and maintain consistent formatting.

pfefferle commented:
I do not think so! I would prefer them to look the same!?

(Domain and Block-Words have a box and white background and the actor list not... At least at the time I wrote the comment 😉)

obenland commented:
I see what you mean!

obenland commented:
@pfefferle I think I'd like to remove the UI for user blocks for now and do that in a follow up PR with a full table etc and make it nice

obenland committed:
Fix test_add_user_block_valid_types by mocking remote actor response

obenland committed:
Remove explicit cap check

obenland committed:
Backslashit

obenland committed:
Remove actor block functionality from moderation settings

This commit removes the ability to block specific ActivityPub actors (users) by their actor URL from both site-wide and user moderation settings. The related UI elements, JavaScript logic, and server-side validation have been updated to only support domain and keyword blocks.

obenland committed:
Merge branch 'trunk' into add/block-lists

obenland committed:
Moderation: Refine UI with singular class names and table-only handling

obenland committed:
Use namespaced functions and improve moderation labels

Replaced calls to array_map and do_action with their namespaced versions for consistency and performance. Updated moderation-related labels and descriptions to use esc_html__ for proper escaping. Renamed Moderation::activity_is_blocked to Moderation::activity_is_blocked_for_user for clarity.

obenland committed:
Refactor moderation to remove actor-level blocking

Actor-level block functionality has been removed from the Moderation class and related tests. Blocking is now only supported for domains and keywords, simplifying the blocklist logic and data structures. Tests and settings have been updated to reflect this change, and duplicate domain entries are now prevented via array_unique in sanitization.

obenland committed:
Merge branch 'trunk' into add/block-lists

obenland commented:
@pfefferle This should be ready for a review

pfefferle commented:
Newly added items will be added twice and even more times and the formatting (even/odd) is only visible after a reload.

<img width="766" height="456" alt="Screenshot 2025-07-30 at 10 43 49" src="https://github.com/user-attachments/assets/3c1e3991-13c2-4462-b461-b1966d010c85" />

Should we validate for a real hos/domain or is it fine to have simple words that will be matched in the domain?

pfefferle commented on includes/class-moderation.php:
```diff
@@ -0,0 +1,251 @@
+		// Extract actor information.
+		$actor_id = '';
+		if ( isset( $activity_data['actor'] ) ) {
+			$actor_id = is_string( $activity_data['actor'] ) ? $activity_data['actor'] : ( $activity_data['actor']['id'] ?? '' );
```
```suggestion
			$actor_id = object_to_uri( $activity_data );
```

pfefferle commented on includes/class-moderation.php:
```diff
@@ -0,0 +1,251 @@
+			if ( $domain && in_array( $domain, $blocked_domains, true ) ) {
+				return true;
+			}
+		}
```
should we fallback to ids and urls of the Activity or the Activity-Object?

obenland replied:
I updated it to include the author Id, activity id and activity object id. I assume the urls use the same domains as ids?

pfefferle replied:
I would assume so, but there is no guarantee.

pfefferle commented:
The current code is very explicit when it comes to domains (only checks the actor-id) but the keywords will be used for the whole `Activity`. Should we be more explicit with the keywords too? Maybe check keywords only in `content` and `summary`?

Should we re-use the Disallowed Comment Keys list for keyword blocking and only add custom code for domains and actors?

pfefferle commented:
We should explain the different options a bit more detailed so that users understand when to use what option and what it will check/block.

Something like:

* "Blocked Domains" – This checks all domains (URLs and IDs) associated with an Activity against the list, but does not examine the content or summary fields.
* "Blocked Keywords" - This checks summaries, contents and usernames, but not Domains.

obenland commented:
> Newly added items will be added twice and even more times and the formatting (even/odd) is only visible after a reload.
> Should we validate for a real hos/domain or is it fine to have simple words that will be matched in the domain?
> —https://github.com/Automattic/wordpress-activitypub/pull/2020#issuecomment-3135390213

Did you test with the latest version of this PR? I can't reproduce that locally, I'm afraid.

pfefferle commented:
> Did you test with the latest version of this PR? I can't reproduce that locally, I'm afraid.

It seems that this only happens if you start with an empty list and add multiple items without reloading. If there are already items in the list, it seems to add the right design.

pfefferle commented:
The "added twice and more" is also "only" a temporary thing. If you reload the site it will be properly uniqueified.

obenland committed:
Fix feedback

Props @pfefferle

obenland committed:
Merge branch 'trunk' into add/block-lists

obenland committed:
Refactor moderation to use Activity objects

Updated Moderation methods and related controllers to accept and process Activity objects instead of arrays. Adjusted tests to use Activity::init_from_array for consistency and improved type safety. This change streamlines activity handling and ensures more reliable moderation logic.

obenland commented:
> Should we be more explicit with the keywords too? Maybe check keywords only in content and summary?

Yup, I like that, updated to only look in content.

> Should we re-use the Disallowed Comment Keys list for keyword blocking and only add custom code for domains and actors?

I worry about fracturing the UI more that it already is between actor and domain & keywords. 
Under the hood we could pipe the keywords into the disallowed comments key list to have `wp_check_comment_disallowed_list()` check for them, too, but I'm not sure whether it's more or less effort.

obenland committed:
Improve ActivityPub moderation JavaScript with duplicate prevention and domain validation

- Add client-side duplicate prevention to prevent submitting already blocked terms
- Add domain validation to ensure proper domain format (requires dot and valid characters)
- Add proper error messages with both accessibility announcements and visible alerts
- Include localized error messages for better user experience
- Prevent invalid domains like "fred" from being submitted

obenland commented:
> We should explain the different options a bit more detailed so that users understand when to use what option and what it will check/block.

I think we're already pretty clear in what they do, not sure end-users should have to worry about details like URLs and IDs.

obenland commented:
@pfefferle It looks like in blog and actor mode, a user-specific block doesn't prevent an activity from going through when the blog actor boosted the user's post. Both get notified of a comment, for example, and the blog actor's notification goes through.

I assume that will be true for any boosts, even between multiple users?

pfefferle commented:
> I think we're already pretty clear in what they do, not sure end-users should have to worry about details like URLs and IDs.

It was not so much about the URLs and IDs, but more what fields will be checked. As an example: The domain block will not search for blocked domains in the content.

obenland commented:
> It was not so much about the URLs and IDs, but more what fields will be checked. As an example: The domain block will not search for blocked domains in the content.

Gotcha! Yeah, I feel like that's an implementation detail we should get right and not something end users should have to consider. It should just work. There'll likely be some iteration based on user feedback, but eventually it should just work.

pfefferle commented:
> Gotcha! Yeah, I feel like that's an implementation detail we should get right and not something end users should have to consider. It should just work. There'll likely be some iteration based on user feedback, but eventually it should just work.

Yes, but it should be also transparent and/or self explanatory. Users should not be confused by why a post might be blocked or not!

mediaformat commented:
> Maybe check keywords only in `content` and `summary`?

I would add Display name to the places keywords should check. Especially since remote comments are not 'editable'.

Abusive/uncivil actors have been known to use slurs in their display names.

pfefferle committed:
add summary and name

pfefferle committed:
Improve object URI handling and keyword checks in moderation

Replaces direct object ID parsing with object_to_uri for more robust URI extraction. Also appends preferred username to content when the object is an actor, enhancing blocked keyword detection.

pfefferle committed:
Merge branch 'trunk' into add/block-lists

pfefferle committed:
Add 'id' field to Note objects in moderation tests

Updated test cases in Test_Moderation to include an 'id' field for Note objects. This ensures test data more accurately reflects expected object structure and improves test reliability.

pfefferle committed:
Add tests for blocked activity attributes in moderation

Introduces a new test method and data provider to verify that activities with certain blocked attributes (such as 'content', 'preferredUsername', 'summary', and 'name') are correctly identified as blocked by the Moderation::activity_is_blocked method.
